x-pack/plugins/security_solution/public/common/hooks/timeline/use_query_timeline_by_id_on_url_change.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useEffect, useMemo } from 'react';
+
+import { useLocation } from 'react-router-dom';
+import usePrevious from 'react-use/lib/usePrevious';
+import { useDispatch } from 'react-redux';
+import type { TimelineUrl } from '../../../timelines/store/timeline/model';
+import { timelineActions, timelineSelectors } from '../../../timelines/store/timeline';
+import { TimelineId, TimelineTabs } from '../../../../common/types';
+import { useShallowEqualSelector } from '../use_selector';
+
+import {
+  dispatchUpdateTimeline,
+  queryTimelineById,
+} from '../../../timelines/components/open_timeline/helpers';
+import {
+  decodeRisonUrlState,
+  getParamFromQueryString,
+  getQueryStringFromLocation,
+} from '../../utils/global_query_string/helpers';
+import { URL_PARAM_KEY } from '../use_url_state';
+
+/**
+ * After the initial load of the security solution, timeline is not updated when the timeline URL search value is changed
+ * This is because those state changes happen in place and don't lead to querying data for the new id.
+ *
+ * But this scenario only happens when there are timelines with conflicted ids. Because it shows a HTML link to the conflicted
+ * timeline, which we do not have control of. We already fetch the data on the onClick callback for all other timeline links in the APP.
+ * For the conflict scenario, we are actively pulling the id changes that take place for the timeline in the URL and calling the query below
+ * to request the new data.
+ *
+ * *** The conflict scenario happens when migrating from an older version to 8.0+
+ */
+export const useQueryTimelineByIdOnUrlChange = () => {
+  const getTimeline = useMemo(() => timelineSelectors.getTimelineByIdSelector(), []);
+  const flyoutTimeline = useShallowEqualSelector((state) => getTimeline(state, TimelineId.active));
+
+  const { search } = useLocation();
+  const oldSearch = usePrevious(search);
+  const timelineIdFromReduxStore = flyoutTimeline?.savedObjectId ?? '';
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    const oldUrlStateString = getQueryStringKeyValue({
```

Comment:
I'm wondering if we could put `oldUrlStateString`, `newUrlStateString`, `newTimeline`, and `oldTimeline` outside of the useEffect, so they have their own dependency. At the same time the dependency of useEffect would be smaller.

Reply:
I can't entirely agree with the overall implementation of this hook. It only exists for an edge case but it runs on every `location.search` change and it parses the search string which is expensive. But your suggestion sounds like a good improvement. I am implementing it 👍